app/src/main/java/org/mozilla/focus/activity/MainMediator.java

```diff
@@ -223,4 +207,75 @@ public void onBackStackChanged() {
             }
         }
     }
+
+    private static class EntryData {
+        /** argument passed to {@link FragmentTransaction#addToBackStack(String)}, pressing back when this
+         * type of fragment is in foreground will close the app */
+        private static final String TYPE_ROOT = "root";
+
+        /** argument passed to {@link FragmentTransaction#addToBackStack(String)}, adding fragment of
+         * this type will make browser fragment go to background */
+        private static final String TYPE_ATTACHED = "attached";
+
+        /** argument passed to {@link FragmentTransaction#addToBackStack(String)}, browsing fragment
+         * will still be in foreground after adding this type of fragment. */
+        private static final String TYPE_FLOATING = "floating";
+
+        String tag;
+        String type;
+    }
+
+    private static class EntryDataSet {
+        private HashMap<String, EntryData> data = new HashMap<>();
+        private HashMap<String, EntryData> backStackRecords = new HashMap<>();
+
+        String add(String tag, String type) {
+            EntryData data = new EntryData();
+            data.tag = tag;
+            data.type = type;
+
+            String key = Integer.toHexString(data.hashCode());
+            this.data.put(key, data);
+
+            logData("addNewEntry");
+            return key;
+        }
+
+        EntryData get(FragmentManager.BackStackEntry entry) {
+            return data.get(entry.getName());
+        }
+
+        void onBackStackChanged(FragmentManager manager) {
```

Comment:
I don't get why we're splitting this into two methods (onBackStackChanged and purge) and why are we maintaining two copies of data. (For concurrency problems?)

Reply:
I'll move the call to purge() into onBackStackChanged()